### DISCUSSION TITLE
Passive Garage Door Opener - state messed up in Home app

### DISCUSSION THREAD

**Discussion by delingren** [HomeSpan/HomeSpan#828] [Q&A]
I'm working on a project to "smartify" a dumb garage opener. There are two parts to this project:

1. Report the state of the door (open, closed, opening, closing, stopped) to Home app.
2. Accept open and close commands from Home app.

Since the door can be controlled manually without using Home app, the state is purely determined by two sensors. The two parts are more or less independent of each other. I am having a little trouble with the first part where it's basically a "passive" garage door opener that only reports the state.

In particular, here is a sequence of events in a scenario:

1. We start with the state being CLOSED, 
1. Then I set it to OPENING (using `CurrentDoorState::setVal()`)
1. A bit later, I set it to OPEN

This basically simulates the actual events that would happen when you open a door. However, in the Home app, when I set the state to OPENING, it shows its state as CLOSING, and it'd stay in that state, even after I set it to OPEN. It'd only change to CLOSED if I set it to CLOSED.

I checked Example 13 which implements my part 2, but not part 1. And in the comment, it mentioned that Home app does some speculation on the intermediary state. However, I couldn't find any mention in the HAP spec. Could I be looking at the wrong spec? Here is what I'm reading: https://github.com/ljezny/Particle-HAP/blob/master/HAP-Specification-Non-Commercial-Version.pdf

And, why would Home app ever transition from CLOSED state to CLOSING state?! It simply doesn't make sense. Any help would be appreciated.

And for anyone who's interested, here's my detailed design and current sketch: https://github.com/delingren/homeautomation/tree/master/garage_door_opener

Thanks.

**Comment by HomeSpan** [comment]
I've just tried Example 13 and it seems to be working as expected.  Can you confirm?  If so, then it is likely something in your sketch - can you post a copy?  

Note you must have a Home Hub (a Home Pod or Apple TV Gen 4) for the sketch to be able to properly report the state of the door.

**Comment by delingren** [comment]
Thanks for your response. Yes, example 13 works fine. What my sketch does that it doesn't do is: I report the state of the door when you operate the door manually, not from the Home app. I use two sensors, one for the open position (upper sensor) and one for the closed position (lower sensor). E.g. when the lower sensor goes from LOW to HIGH, we know the door is being opened and I set the door state to OPENING, but the Home app thinks its state is CLOSING.

[Here](https://github.com/delingren/homeautomation/blob/master/garage_door_opener/garage_door_opener.ino) is a full list of my sketch. Sorry it's a bit long, due to the nature of the state machine logic. But it's reasonably well documented and I have a detailed [design doc](https://github.com/delingren/homeautomation/blob/master/garage_door_opener/README.md).

I'm also pasting a shortened version here, with the obstruction detection logic removed:
```
#include "HomeSpan.h"
#include <Arduino.h>
#include <vector>

struct DEV_GarageDoor : Service::GarageDoorOpener {
  // Debounce time for the contact sensors. Adjust based the particular sensor
  // and probably the speed of the motor.
  const unsigned long millis_debounce = 600;

  // These values are the same as the enum defined in
  // GarageDoorOpener::CurrentDoorState. We would use that enum but it's
  // anonymous.
  enum door_state {
    OPEN = 0,
    CLOSED = 1,
    OPENING = 2,
    CLOSING = 3,
    STOPPED = 4
  };

  static std::vector<DEV_GarageDoor *> all_openers;

  static void isr() {
    // The interrupt handler for all sensor pins. I don't think there is an easy
    // way to create an ISR for each individual pin. So they'll have to share an
    // ISR.
    for (auto it = all_openers.begin(); it != all_openers.end(); it++) {
      uint8_t new_value_upper = digitalRead((*it)->pin_upper);
      uint8_t new_value_lower = digitalRead((*it)->pin_lower);

      // loop() will do the debouncing logic. We are only informing it that one
      // of the sensors has changed state and when.
      if (new_value_lower != (*it)->value_lower ||
          new_value_upper != (*it)->value_upper) {
        (*it)->sensor_triggered = true;
        (*it)->millis_last = millis();
      }
    }
  }

  Characteristic::CurrentDoorState current_state;
  Characteristic::TargetDoorState target_state;
  Characteristic::ObstructionDetected obstruction;

  uint8_t pin_upper;
  uint8_t pin_lower;
  uint8_t pin_relay;

  uint8_t value_upper;
  uint8_t value_lower;

  door_state state = door_state::STOPPED;

  bool sensor_triggered = false;
  unsigned long millis_last = 0;

  DEV_GarageDoor(uint8_t pin_upper, uint8_t pin_lower, uint8_t pin_relay)
      : Service::GarageDoorOpener() {
    this->pin_upper = pin_upper;
    this->pin_lower = pin_lower;
    this->pin_relay = pin_relay;

    pinMode(pin_upper, INPUT_PULLUP);
    pinMode(pin_lower, INPUT_PULLUP);

    value_upper = digitalRead(pin_upper);
    value_lower = digitalRead(pin_lower);

    pinMode(pin_relay, OUTPUT);
    digitalWrite(pin_relay, LOW);

    // Determine the initial state.
    if (value_upper == LOW && value_lower == HIGH) {
      state = door_state::OPEN;
    } else if (value_upper == HIGH && value_lower == LOW) {
      state = door_state::CLOSED;
    } else {
      // The door could be OPENING, CLOSING, or STOPPED. But there is no way we
      // can tell. So we assume it's STOPPED. If it's indeed OPENING or CLOSING,
      // it'll trigger a sensor change event (if successful) and we will be able
      // to transition into the OPEN or CLOSED state. If unsuccessful, STOPPED
      // will be the correct state.
      state = door_state::STOPPED;
    }
    current_state.setVal(state);

    all_openers.push_back(this);
    attachInterrupt(pin_upper, isr, CHANGE);
    attachInterrupt(pin_lower, isr, CHANGE);
  }

  static unsigned long millis_since(unsigned long millis_then) {
    unsigned long millis_now = millis();
    // millis() overflows after about 50 days. So, if millis_now < millis_last,
    // that's what happened.
    return millis_now >= millis_then
               ? millis_now - millis_then
               : 0xFFFFFFFF - (millis_then - millis_now) + 1;
  }

  void transition(door_state new_state) {
    Serial.printf("-- Transitioning from %d to %d.\n", state, new_state);

    state = new_state;
    current_state.setVal(state);
  }

  boolean update() { return true; }

  void loop() {
    // State transitions.
    //         |   UR   |   UF   |   LR   |   LF   |   TO   |
    // --------+--------+--------+--------+--------+--------+
    // open    |closing |        |        |        |        |
    // closed  |        |        |opening |        |        |
    // opening |        |  open  |        | closed |stopped |
    // closing |        |  open  |        | closed |stopped |
    // stopped |        |  open  |        | closed |        |
    // --------+--------+--------+--------+--------+--------+

    // All empty cells are invalid. However, we always transition the the state
    // determined by the previous pin values and current ones, even if we are in
    // an invalid state. It's the only way we can recover. But we report such
    // errors.

    // Sensor change event. This happens when the door reaches or leaves either
    // end of the track. It's not practically possible to have two transitions
    // in one iteration, since it takes at least a few seconds to open or close
    // a garage door. hence we only consider one transition here.
    if (sensor_triggered && millis_since(millis_last) > millis_debounce) {
      // Debouncing logic: it's been enough time since the last change, so it
      // should be stable now.
      uint8_t new_value_upper = digitalRead(pin_upper);
      uint8_t new_value_lower = digitalRead(pin_lower);

      if (value_upper == LOW && new_value_upper == HIGH) {
        // Upper, Rising
        if (state != door_state::OPEN) {
          Serial.printf("-- Upper sensor falling. Invalid state: %d.\n", state);
        }
        transition(door_state::CLOSING);
      } else if (value_upper == HIGH && new_value_upper == LOW) {
        // Upper, Falling
        if (state != door_state::OPENING && state != door_state::CLOSING &&
            state != door_state::STOPPED) {
          Serial.printf("-- Upper sensor rising. Invalid state: %d.\n", state);
        }
        transition(door_state::OPEN);
      } else if (value_lower == LOW && new_value_lower == HIGH) {
        // Lower, Rising
        if (state != door_state::CLOSED) {
          Serial.printf("-- Lower sensor falling. Invalid state: %d.\n", state);
        }
        transition(door_state::OPENING);
      } else if (value_lower == HIGH && new_value_lower == LOW) {
        // Lower, Falling
        if (state != door_state::OPENING && state != door_state::CLOSING &&
            state != door_state::STOPPED) {
          Serial.printf("-- Lower sensor rising. Invalid state: %d. \n", state);
        }
        transition(door_state::CLOSED);
      }

      value_upper = new_value_upper;
      value_lower = new_value_lower;
      sensor_triggered = false;
    }
  }
};

std::vector<DEV_GarageDoor *> DEV_GarageDoor::all_openers;

void setup() {
  Serial.begin(115200);

  homeSpan.setStatusPin(32).setControlPin(26).begin(Category::GarageDoorOpeners,
                                                    "Garage Door");

  new SpanAccessory();
  new Service::AccessoryInformation();
  new Characteristic::Identify();
  new Characteristic::Name("Garage Door");
  new DEV_GarageDoor(23, 22, 13);

  homeSpan.autoPoll();
}

void loop() {}
```

**Comment by delingren** [comment]
Additional observation: in the scenario that I described (the door transitions from CLOSED to OPENING), if I try to open it in the Home app before the events, even if I am *not* doing anything in `update()`, the tile would correctly transition from CLOSED to OPENING. I.e, the following is what happens, and this is correct.

1. The door is CLOSED. My sketch has called `setVal(CLOSED)`, and the app correctly shows the state.
1. In the Home app, I try to close the door. But my sketch does nothing whatsoever in respond to the close command.
1. The tile in the Home app shows the door as OPENING.
1. My sketch sets the state to OPENING.
1. A few seconds later, my sketch sets the state to OPEN.
1. The tile in the Home app shows the door as OPEN. 

In contrast, the previous scenario is *incorrect*:

1. The door is CLOSED. My sketch has called `setVal(CLOSED)`, and the app correctly shows the state.
1. My sketch sets the state to OPENING.
1. The tile in the Home app shows the door as *CLOSING*
1. A few seconds later, my sketch sets the state to OPEN.
1. The tile in the Home app *still* shows the door as CLOSING.
1. Now, my sketch sets the state to CLOSING, then CLOSED a few seconds later.
1. The tile in the Home app now shows CLOSED.

**Reply by HomeSpan** [reply]
If you manually operate the garage door you need to tell the Home App what the new desired state is. Updating just the current state tells the Home App that the door may be opening, in spite of the fact that the Home App thinks the user wants the door closed. 

If you operate the door manually you need to update the target state so that the Home App stays in sync.  You should not have to change the current state as this should be updated by whatever logic would do so when target state changes. 

In other words: If the Home App changes the target state, the sketch should respond by changing current state.  If the door is operated manually, you should change the target state, and the rest follows from above. 

A few other thoughts:  For the contact sensors, you can use the HomeSpan SpanButton class. It includes debouncing and integrates nicely with Services classes. No need to use interrupts for something like this.

**Comment by delingren** [comment]
Huh. Thanks for the explanation. That logic is a little interesting. I thought target state would be read only. After adding that logic, everything seems to be working correctly, at least on the breadboard.
Also, thanks for the pointer to SpanButton. I ended up using SpanToggle since I want to act on both rising and falling edges.
